File: ProblemSolving/ProblemSolving/Programmers/PG81301_숫자문자열과영단어.cpp
#include <cstdio>
#include <string>
#include <vector>
using namespace std;

int trans(string s, int* idx) {
    
    int num = 0;

    switch (s[*idx]) {
        case 'z':
            *idx += 4;
            num = 0;
            break;
        case 'o':
            *idx += 3;
            num = 1;
            break;
        case 't':
            if (s[*idx + 1] == 'w') {
                *idx += 3;
                num = 2;
            }
            else {
                *idx += 5;
                num = 3;
            }
            break;
        case 'f':
            if (s[*idx + 1] == 'o') {
                *idx += 4;
                num = 4;
            }
            else {
                *idx += 4;
                num = 5;
            }
            break;
        case 's':
            if (s[*idx + 1] == 'i') {
                *idx += 3;
                num = 6;
            }
            else {
                *idx += 5;
                num = 7;
            }
            break;
        case 'e':
            *idx += 5;
            num = 8;
            break;
        case 'n':
            *idx += 4;
            num = 9;
            break;
    }

    return num;
}

int solution(string s) {
    int answer = 0;

    for (int i = 0; i < s.size(); ) {
        if (s[i] >= '0' && s[i] <= '9') {
            answer = answer * 10 + (s[i] - '0');
            i++;
        }
        else {
            int num = trans(s, &i);
            answer = answer * 10 + num;
        }
    }

    return answer;
}

int main() {

    string s = "one4seveneight";
    printf("%d", solution(s));

    return 0;
}

File: ProblemSolving_Java/ProblemSolving_Java/src/programmers/PG81301_숫자문자열과영단어.java
package programmers;

public class PG81301_숫자문자열과영단어 {

    public static void main(String[] args) {

        String s = "one4seveneight";

        Solution solution = new Solution();

        int answer = solution.solution(s);
        System.out.println("answer = " + answer);
    }

    static class Solution {
        public int solution(String s) {
            int answer = 0;

            // 숫자면 스킵
            // 단어면 숫자까지 잘라서
                // 어떤 숫자인지 확인하고 추가

            String answerStr = "";

            int sIndex = 0;
            String tmpStr = "";
            while(sIndex < s.length()) {
                char currentChar = s.charAt(sIndex);

                if(isNumber(currentChar))
                    answerStr+=currentChar;
                else {
                    tmpStr += currentChar;

                    String numberStr = getNumberStr(tmpStr);

                    if(!numberStr.equals("X")) {
                        answerStr += numberStr;
                        tmpStr = "";
                    }
                }

                sIndex++;
            }

            answer = Integer.parseInt(answerStr);

            return answer;
        }

        private String getNumberStr(String tmpStr) {
            if(tmpStr.equals("zero"))
                return "0";
            else if(tmpStr.equals("one"))
                return "1";
            else if(tmpStr.equals("two"))
                return "2";
            else if(tmpStr.equals("three"))
                return "3";
            else if(tmpStr.equals("four"))
                return "4";
            else if(tmpStr.equals("five"))
                return "5";
            else if(tmpStr.equals("six"))
                return "6";
            else if(tmpStr.equals("seven"))
                return "7";
            else if(tmpStr.equals("eight"))
                return "8";
            else if(tmpStr.equals("nine"))
                return "9";
            else
                return "X";
        }

        private boolean isNumber(char charAt) {
            if(charAt >= '0' && charAt <= '9')
                return true;
            return false;
        }
    }
}


